dataset/shape.cpp

```diff
@@ -141,4 +141,271 @@ Dataset resize(const DatasetConstView &d, const Dim dim,
   return result;
 }
 
+namespace {
+
+/// Split dims for reshaping one dim into multiple dims
+///
+/// Go through the old dims and:
+/// - if the dim does not equal the dim that is being stacked, copy dim/shape
+/// - if the dim equals the dim to be stacked, replace by stack of new dims
+const Dimensions split_dims(const Dimensions &old_dims, const Dim from_dim,
+                            const Dimensions &to_dims) {
+  Dimensions new_dims;
+  for (const auto dim : old_dims.labels())
+    if (dim != from_dim)
+      new_dims.addInner(dim, old_dims[dim]);
+    else
+      for (const auto lab : to_dims.labels())
+        new_dims.addInner(lab, to_dims[lab]);
+  return new_dims;
+}
+
+/// Flatten dims for reshaping multiple dims into one
+///
+/// Go through the old dims and:
+/// - if the dim is contained in the list of dims to be flattened, add the new
+///   dim once
+/// - if not, copy the dim/shape
+const Dimensions flatten_dims(const Dimensions &old_dims,
+                              const Dimensions &from_dims, const Dim to_dim) {
+  Dimensions new_dims;
+  for (const auto dim : old_dims.labels())
+    if (from_dims.contains(dim)) {
+      if (!new_dims.contains(to_dim))
+        new_dims.addInner(to_dim, from_dims.volume());
```

Comment:
What is happening here regarding dimension order is quite important and missing from the docs for `flatten`. Note that https://numpy.org/doc/stable/reference/generated/numpy.ndarray.flatten.html allows dimension order to be specified. If we do not allow this (could be achieved with a reshape first), we at least need to explain what is done about fastest dimensions.

Reply:
hmm, we don't support specifying the order for reshape, so I didn't support it or document it here either.
I'm note sure doing a `reshape` here would help with supporting order? Shouldn't it be a `transpose` instead?

Reply:
My point is, it might not be clear to users (undocumented) how this will behave.  Taking your example from the docs:
```python
N = 6
M = 4
x = sc.array(dims=['x'], values=np.arange(N))
y = sc.array(dims=['y'], values=np.arange(M))
xy = sc.array(dims=['y', 'x'], values=np.arange(N*M).reshape([M,N]))
data = sc.DataArray(data=xy, coords={'xx':x,'yy':y})
sc.flatten(data, dim='p').coords['xx'].values
```
But with x,y transposed
```python
xy = sc.array(dims=['x', 'y'], values=np.arange(N*M).reshape([N,M]))
data = sc.DataArray(data=xy, coords={'xx':x,'yy':y})
sc.flatten(data, dim='p').coords['xx'].values
```
Makes sense if you know that you are appending along inner dimension.